db/db_impl/db_impl_open.cc

```diff
@@ -286,6 +286,12 @@ Status DBImpl::ValidateOptions(const DBOptions& db_options) {
         "atomic_flush is currently incompatible with best-efforts recovery");
   }
 
+  if (db_options.use_direct_io_for_flush_and_compaction &&
+      0 == db_options.writable_file_max_buffer_size) {
```

Comment:
If the underlying FileSystem does not support direct io, should this still return a failure or is that caught somewhere else?

Reply:
I think opening a file in direct mode on a platform lacking direct IO support will return non-ok status. Furthermore, we perform a check at db open: https://github.com/facebook/rocksdb/blob/6.28.fb/db/db_impl/db_impl_open.cc#L475:L484.

Back to this comment. I think we should return early on error and give specific error message.

Reply:
So why not either 1) do something similar to the Open in this case or 2) (even better) move the check into a FileSystem::SupportsDirectIO method?

Reply:
For 1), if Open already tests support for direct IO, why do it here? The purpose of the check is to inform users that the combination is wrong.

For 2), do you mean moving the check from `Open` to `FileSystem`? This can be done, but is beyond the scope of this PR.